File: backend/app/schemas/results.py
"""Pydantic response schemas for the results / season-scoreboard read API.

Mirrors the Pydantic v2 ``BaseModel`` + ``ConfigDict(extra="forbid")`` style of
:mod:`app.schemas.picks`. Each schema is built explicitly from the
:mod:`app.services.standings` service objects via a ``from_*`` classmethod
(mirroring :meth:`app.schemas.picks.PickRead.from_orm_pick`) rather than coupling
to ORM rows.

Privacy posture: like ``PickRead``, these expose ``display_name`` only — never
``user_id``. The scoreboard is shared-among-members but still does not leak the
internal user surrogate id.
"""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict

from app.demo.oracle import Standings
from app.models import PickType
from app.schemas.types import DiscordId
from app.services.standings import UserIdentity
from app.services.standings import UserWeekResult as UserWeekResultService
from app.services.standings import WeekResultPick


class WeekResultPickRead(BaseModel):
    """One graded pick within a user's week."""

    model_config = ConfigDict(extra="forbid")

    game_id: int
    pick_type: PickType
    is_mortal_lock: bool
    outcome: str  # the GradeOutcome string value
    points: int
    # MISC free-text prediction; only ever present on a revealed entry (the
    # service privacy gate hides an other-user pick on a not-yet-locked game).
    misc_text: str | None = None

    @classmethod
    def from_service(cls, pick: WeekResultPick) -> "WeekResultPickRead":
        """Build from a service :class:`~app.services.standings.WeekResultPick`."""
        return cls(
            game_id=pick.game_id,
            pick_type=pick.pick_type,
            is_mortal_lock=pick.is_mortal_lock,
            outcome=pick.outcome,
            points=pick.points,
            misc_text=pick.misc_text,
        )


class UserWeekResult(BaseModel):
    """One user's graded picks + weekly score for a ``{season, week}``.

    ``user_id`` is deliberately absent (display_name only).
    """

    model_config = ConfigDict(extra="forbid")

    display_name: str
    weekly_score: int
    picks: list[WeekResultPickRead]
    # Avatar identity (mirrors the service UserWeekResult). Both None when the
    # user has no Discord avatar; the frontend builds the CDN URL from these and
    # falls back to initials otherwise. user_id stays deliberately absent.
    discord_id: DiscordId = None
    discord_avatar_hash: str | None = None

    @classmethod
    def from_service(cls, result: UserWeekResultService) -> "UserWeekResult":
        """Build from a service :class:`~app.services.standings.UserWeekResult`."""
        return cls(
            display_name=result.display_name,
            weekly_score=result.weekly_score,
            picks=[WeekResultPickRead.from_service(p) for p in result.picks],
            discord_id=result.discord_id,
            discord_avatar_hash=result.discord_avatar_hash,
        )


class WeekResultsResponse(BaseModel):
    """Per-week graded results for a ``{season, week}`` across all users."""

    model_config = ConfigDict(extra="forbid")

    season: int
    week: int
    results: list[UserWeekResult]

    @classmethod
    def from_results(
        cls, *, season: int, week: int, results: list[UserWeekResultService]
    ) -> "WeekResultsResponse":
        """Shape the service per-user results into the HTTP response."""
        return cls(
            season=season,
            week=week,
            results=[UserWeekResult.from_service(r) for r in results],
        )


class SeasonStandingRow(BaseModel):
    """One user's cumulative season standing row.

    ``weekly_scores`` maps week number -> that week's integer score.
    """

    model_config = ConfigDict(extra="forbid")

    display_name: str
    season_total: int
    weekly_scores: dict[int, int]
    # Avatar identity, joined by the unique display_name from the standings
    # service's identity map. Both None when the user has no Discord avatar; the
    # frontend builds the CDN URL from these and falls back to initials.
    discord_id: DiscordId = None
    discord_avatar_hash: str | None = None


class SeasonStandingsResponse(BaseModel):
    """Cumulative season standings over all users, pre-ordered by the service."""

    model_config = ConfigDict(extra="forbid")

    season: int
    standings: list[SeasonStandingRow]
    # True ONLY when the season has games and EVERY game is FINAL — the
    # season-end state that the frontend uses to award 1st/2nd/3rd medals. False
    # for an in-progress season and for a season with zero games.
    season_complete: bool

    @classmethod
    def from_standings(
        cls,
        *,
        season: int,
        standings: Standings,
        season_complete: bool,
        identities: dict[str, UserIdentity] | None = None,
    ) -> "SeasonStandingsResponse":
        """Shape the service :class:`~app.demo.oracle.Standings` into the response.

        Preserves the service's ``(-season_total, display_name)`` ordering (the
        rows are emitted in the order the service produced them). ``season_complete``
        is the route-computed season-end flag (see
        :func:`app.services.standings.season_is_complete`).

        ``identities`` maps each row's unique ``display_name`` to its
        :class:`~app.services.standings.UserIdentity` (``discord_id`` +
        ``discord_avatar_hash``). A missing/absent entry yields ``None`` for both
        avatar fields — never raises (defensive: a row without an identity simply
        renders initials downstream).
        """
        identities = identities or {}
        return cls(
            season=season,
            season_complete=season_complete,
            standings=[
                SeasonStandingRow(
                    display_name=row.display_name,
                    season_total=row.season_total,
                    weekly_scores=dict(row.weekly_scores),
                    discord_id=(
                        identities[row.display_name].discord_id
                        if row.display_name in identities
                        else None
                    ),
                    discord_avatar_hash=(
                        identities[row.display_name].discord_avatar_hash
                        if row.display_name in identities
                        else None
                    ),
                )
                for row in standings.results
            ],
        )

File: frontend/src/lib/results.ts
/**
 * Typed client for the /api/results read endpoints (season scoreboard + weekly).
 *
 * Mirrors the small, one-function-per-endpoint style of lib/picks.ts: every call
 * routes through the CSRF-aware `api<T>()` wrapper (credentials:include +
 * X-CSRF-Token on unsafe methods) — never a raw fetch. These are read-only GETs,
 * so no CSRF token is attached, but they still go through api() for the
 * credentialed fetch + ApiError handling.
 *
 * Wire shapes match the backend exactly (see backend/app/schemas/results.py).
 * The Weekly screen will later extend this file with `getWeekResults`.
 */
import { api } from "./api";
import type { PickType } from "./picks";

/**
 * One graded pick within a user's week (mirrors backend
 * WeekResultPickRead in backend/app/schemas/results.py). `outcome` is the
 * GradeOutcome string value: WIN | LOSS | PUSH | INELIGIBLE | UNGRADEABLE
 * (UNGRADEABLE is the neutral "not yet scored" state for not-yet-FINAL games).
 * `pick_type` reuses the PickType union from ./picks (do NOT redefine it).
 */
export interface WeekResultPickRead {
  game_id: number;
  pick_type: PickType;
  is_mortal_lock: boolean;
  outcome: string;
  points: number;
  /**
   * The free-text prediction for a MISC pick (mirrors backend
   * WeekResultPickRead.misc_text). Present ONLY on a REVEALED entry — the server
   * already redacts other users' MISC text until the tied game locks, so an
   * unrevealed/other-user MISC simply arrives with this null/absent (or the whole
   * entry omitted). Null/absent for every non-MISC pick.
   */
  misc_text?: string | null;
}

/**
 * One user's graded picks + weekly score for a {season, week} (mirrors backend
 * UserWeekResult). `user_id` is deliberately ABSENT — display_name only.
 *
 * NOTE: OTHER users' picks on not-yet-locked games are OMITTED server-side
 * (the leak gate), so `picks` may contain fewer entries than the user actually
 * made; `weekly_score` is always whole.
 */
export interface UserWeekResult {
  display_name: string;
  weekly_score: number;
  picks: WeekResultPickRead[];
  /**
   * Discord snowflake id for this user as a STRING (mirrors the backend
   * UserWeekResult additions; a 64-bit id loses precision as a JSON number).
   * Null when the user has no Discord identity; paired with
   * `discord_avatar_hash` to build the avatar CDN URL (else initials fallback).
   */
  discord_id: string | null;
  /**
   * Discord avatar hash (mirrors the backend UserWeekResult addition). Null when
   * the user has no custom avatar; the `<Avatar>` component falls back to
   * initials when this (or `discord_id`) is null.
   */
  discord_avatar_hash: string | null;
}

/**
 * Per-week graded results across all users (mirrors backend
 * WeekResultsResponse). `results` is PRE-ORDERED by the backend
 * (`(-weekly_score, display_name)`) — render in this order, do NOT re-sort.
 */
export interface WeekResultsResponse {
  season: number;
  week: number;
  results: UserWeekResult[];
}

/** Fetch the per-week graded results (all users) for a {season, week}. */
export function getWeekResults(
  season: number,
  week: number,
): Promise<WeekResultsResponse> {
  return api<WeekResultsResponse>(
    `/api/results/week?season=${season}&week=${week}`,
  );
}

/**
 * One user's cumulative season standing row.
 *
 * `weekly_scores` maps a week number to that week's integer score. The KEYS are
 * STRINGS, not numbers: the backend type is `dict[int, int]`
 * (backend/app/schemas/results.py SeasonStandingRow), but JSON object keys are
 * always strings, so week numbers serialize as string keys (`"1"`, `"2"`, …).
 * Read a cell with `weekly_scores[String(week)]`.
 */
export interface SeasonStandingRow {
  display_name: string;
  season_total: number;
  weekly_scores: Record<string, number>;
  /**
   * Discord snowflake id for this user as a STRING (mirrors the backend
   * SeasonStandingRow additions; a 64-bit id loses precision as a JSON number).
   * Null when the user has no Discord identity; paired with
   * `discord_avatar_hash` to build the avatar CDN URL (else initials fallback).
   */
  discord_id: string | null;
  /**
   * Discord avatar hash (mirrors the backend SeasonStandingRow addition). Null
   * when the user has no custom avatar; the `<Avatar>` component falls back to
   * initials when this (or `discord_id`) is null.
   */
  discord_avatar_hash: string | null;
}

/**
 * Cumulative season standings over all users, PRE-ORDERED by the backend
 * service (`(-season_total, display_name)`) — render rows in this exact order;
 * do not re-sort client-side.
 */
export interface SeasonStandingsResponse {
  season: number;
  standings: SeasonStandingRow[];
  /**
   * True ONLY when every game in the season is FINAL (mirrors the backend
   * SeasonStandingsResponse.season_complete). The Standings page uses this to
   * award 1st/2nd/3rd medals; it is false for an in-progress season and for a
   * season with zero games.
   */
  season_complete: boolean;
}

/** Fetch the season scoreboard (one row per player) for a season. */
export function getStandings(season: number): Promise<SeasonStandingsResponse> {
  return api<SeasonStandingsResponse>(
    `/api/results/standings?season=${season}`,
  );
}
